sys/patches/not_safe_for_wasi.patch

```diff
@@ -0,0 +1,31 @@
+diff --git a/quickjs.c b/quickjs.c
+index 48aeffc..cb6c80d 100644
+--- a/quickjs.c
++++ b/quickjs.c
+@@ -38,6 +38,8 @@
+ #include <malloc.h>
+ #elif defined(__FreeBSD__)
+ #include <malloc_np.h>
++#else
++#include <malloc.h>
+ #endif
+ 
+ #include "cutils.h"
+@@ -62,17 +64,6 @@
+ #define MALLOC_OVERHEAD  8
+ #endif
+ 
+-#if !defined(_WIN32)
+-/* define it if printf uses the RNDN rounding mode instead of RNDNA */
+-#define CONFIG_PRINTF_RNDN
+-#endif
+-
+-/* define to include Atomics.* operations which depend on the OS
+-   threads */
+-#if !defined(EMSCRIPTEN)
+-#define CONFIG_ATOMICS
+-#endif
```

Comment:
Could we use WASI-specific cpp gate to pass through that instead?
For example:
```diff
-#if !defined(_WIN32)
+#if !defined(_WIN32) && !defined(__wasi__)
 /* define it if printf uses the RNDN rounding mode instead of RNDNA */ 
 #define CONFIG_PRINTF_RNDN
 #endif

 /* define to include Atomics.* operations which depend on the OS
    threads */
-#if !defined(EMSCRIPTEN)
+#if !defined(EMSCRIPTEN) && !defined(__wasi__)
 #define CONFIG_ATOMICS
 #endif
```

I'm not familiar with WASI in C so I don't know that the correct predefined macro we should use, but seems `__wasi__` is a right choice.

Reply:
Nitpick: it might be better to use the standard libc macro `#ifdef _REENTRANT`.

It's:

 - defined by regular native compilers as they build with threads
 - not defined by Emscripten & wasi-sdk clang by default as they don't support threads
 - **and** will allow to correctly detect `emcc -pthread` - that is, if someone is building a multithreaded Wasm build, then it will enable atomics as it should (something it currently doesn't)

Reply:
Btw one update here is that WASI SDK now does have experimental thread support, so using correct condition is even more important: https://github.com/WebAssembly/wasi-sdk/releases/tag/wasi-sdk-20%2Bthreads